extensions/s3/src/main/java/io/deephaven/extensions/s3/ModuloBasedRequestCache.java

```diff
@@ -0,0 +1,75 @@
+//
+// Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+//
+package io.deephaven.extensions.s3;
+
+import io.deephaven.hash.KeyedLongObjectHashMap;
+import io.deephaven.hash.KeyedLongObjectKey;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import software.amazon.awssdk.services.s3.S3Uri;
+import io.deephaven.extensions.s3.S3ChannelContext.Request;
+
+import static java.lang.Math.abs;
+
+/**
+ * This class uses a modulo-based mapping function with a {@link KeyedLongObjectHashMap} to cache {@link Request}
+ * objects.
+ */
+final class ModuloBasedRequestCache implements S3RequestCache {
+
+    private final class RequestIdKey extends KeyedLongObjectKey.BasicStrict<Request> {
+        @Override
+        public long getLongKey(@NotNull final Request request) {
+            return cacheIndex(request.getUri(), request.getFragmentIndex());
+        }
+    }
+
+    private final int maxSize;
+    private final KeyedLongObjectHashMap<Request> requests;
```

Comment:
Given mod, if we are going w/ KeyedXMap, this would be better as KeyedIntegerObjectHashMap.

That said, I do question this approach. Moving to a shared cache, I wonder if we would do better with a more "proper" cache that takes into account the full key (S3Uri + fragmentIndex).

https://github.com/google/guava/wiki/CachesExplained
https://github.com/ben-manes/caffeine

I worry that this "cache" is going to easily get stomped on

Reply:
I made the change to KeyedIntegerObjectHashMap, thanks for the suggestion.
For the caching logic, my idea was to keep something light weight since this cache will primarily only be beneficial for smaller files which have fewer fragments and therefore have lesser probability of being stomped on.